clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1004,6 +1010,26 @@ public ListOffsetResult() {
         return subscriptions.fetchablePartitions(tp -> !exclude.contains(tp));
     }
 
+    /**
+     * Determine which replica to read from.
+     */
+    private Node selectReadReplica(TopicPartition partition, Node leaderReplica) {
+        Optional<Node> node = subscriptions.preferredReadReplica(partition, time.milliseconds())
+                .flatMap(nodeId -> Optional.ofNullable(metadata.fetch().nodeById(nodeId)));
+        if (node.isPresent()) {
+            if (Arrays.asList(metadata.fetch().partition(partition).offlineReplicas()).contains(node.get())) {
```

Comment:
Would it make sense to have a single API which gets the Node for a partition if it is not offline? For example:
```java
Optional<Node> nodeIfOnline(int nodeId, TopicPartition partition, int nodeId);
```

Reply:
Only issue then is we can't know when a node is offline or just non-existent.